### PR TITLE
Add source_tree implementation and tests

### PR DESCRIPTION
**Pull request by abaptiste**
Add the `source_tree` module that abstracts git operations to manipulate and deduce information about a source tree.

Unit tests for the source tree module are included in this diff.

Signed-off-by: abaptiste <abaptiste@users.noreply.github.com>

cc: @mum4k , @landesherr

### CODE REVIEW

File: salvo/src/lib/test_source_tree.py [hunk 1]
```diff
@@ -9,83 +9,240 @@
 from src.lib import (cmd_exec, source_tree)
 import api.source_pb2 as proto_source
 
+logging.basicConfig(level=logging.DEBUG)
 log = logging.getLogger(__name__)
 
 def test_source_tree_object():
   """Verify that we throw an exception if not all required data is present."""
-  pass
+  st = source_tree.SourceTree(proto_source.SourceRepository())
```
Comment (mum4k) [inline salvo/src/lib/test_source_tree.py:17]: The Python style guide advises to use descriptive names, even if they are slightly longer. Can we use a more descriptive identifier instead of just `st`? Consider `source` or `tree`.

Applies to multiple locations below.

File: salvo/src/lib/test_source_tree.py [hunk 2]
```diff
@@ -9,83 +9,240 @@
...
 def _generate_source_tree_from_origin(origin):
   """Build a source tree from a remote url."""
-  pass
+  source_repository = proto_source.SourceRepository(
+      source_path='/tmp',
+      source_url=origin
+  )
+  return source_tree.SourceTree(source_repository)
 
 def mock_run_command_side_effect(*args):
```
Comment (mum4k) [inline salvo/src/lib/test_source_tree.py:106]: Since this is a test helper, future maintainers will need to understand it. Can we unwrap the `*args` into an exact list of arguments and document then and their behavior in the `Args` section of the docstring here?

Applies to all such test helpers below.

File: salvo/src/lib/test_source_tree.py [hunk 3]
```diff
@@ -9,83 +9,240 @@
...
-  pass
+  """
+    Verify that we can identify one commit prior to a specified hash.
+    """
+  origin = 'https://github.com/someawesomeproject/repo.git'
+  st = _generate_source_tree_from_origin(origin)
+
+  commit_hash = 'fake_commit_hash_1'
+
+  mock_check_output.side_effect = _check_output_side_effect
```
Comment (mum4k) [inline salvo/src/lib/test_source_tree.py:243]: A question on complexity - do we need to define `_check_output_side_effect()` as a separate function which combines all possible outputs? This makes the reader scroll up and down from the test case to the function and requires the reader to understand the side_effect function.

Can we just set the return value here in the test so the reader has everything in one location? E.g. just set `mock_check_output.return_value`.

Seems like that would allow us to remove some of the helper code and keep test cases self-contained.

Applies to `_check_output_side_effect_fail` as well.

File: salvo/src/lib/test_source_tree.py [hunk 4]
```diff
@@ -9,83 +9,240 @@
...
+    magic_mock.assert_called_once_with("git remote -v", cmd_params)
 
 def test_get_origin_ssh():
   """Verify that we can determine the origin for a local repository.
 
   In this instance the repo was cloned via ssh.
   """
-  pass
+  remote_string = 'origin  git@github.com:username/reponame.git (fetch)'
+  gitcmd = "git remote -v"
```
Comment (mum4k) [inline salvo/src/lib/test_source_tree.py:60]: (nit) The name should be `git_cmd` since these are two words.

Applies to multiple locations.

File: salvo/src/lib/test_source_tree.py [hunk 5]
```diff
@@ -94,50 +251,321 @@ def _check_output_side_effect_fail(*args):
...
   """Verify that we can list tags from a repository."""
-  pass
+
+  origin = 'https://github.com/someawesomeproject/repo.git'
+  st = _generate_source_tree_from_origin(origin)
+
+  gitcmd = "git tag --list --sort v:refname"
+  git_output = """v1.0.0
+v1.1.0
+v1.2.0
```
Comment (mum4k) [inline salvo/src/lib/test_source_tree.py:363]: Test data are also code in the sense that they need to be understood and maintained with the code changes. Can we minimize this test data (the list of tags) to the bare minimum necessary to test the behavior?

This will decrease the amount of lines we have to maintain here and make it easier to understand the purpose of the test. It isn't clear at the moment if all of these are needed and / or some of them test some special behaviors.

Applies to the remaining two tag lists below as well.

**Review by mum4k — CHANGES_REQUESTED** [review]

File: salvo/src/lib/common/BUILD [hunk 6]
```diff
@@ -0,0 +1,12 @@
+load("@rules_python//python:defs.bzl", "py_library")
+
+package(
+  default_visibility = ["//:__subpackages__"],
+)
+
+py_library(
+    name = "fileops",
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/common/BUILD:8]: Certainly.  I will update.

File: salvo/src/lib/common/fileops.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      to read only and is an optional argument.
+  """
+  data = {}
+  with open(path, mode) as json_file:
+    data = json.loads(json_file.read())
+  return data
+
+
+def open_yaml(path: str, mode: str = 'r') -> dict:
+  """Open a json file and return its contents as a dictionary
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/common/fileops.py:28]: Ack!  Fixed.

File: salvo/src/lib/common/fileops.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+  Returns:
+    The full pathname of the temporary directory.
+  """
+
+  if not os.path.exists(path):
+    os.mkdir(path)
+
+  temp_dir = tempfile.TemporaryDirectory(dir=path)
+  return temp_dir.name
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/common/fileops.py:68]: Yeah, I agree, though the directory lifespan management is something I need to focus on a bit more down the road.  Let me create an issue for this and revisit it (soon) since this change reverberates through the rest of Salvo.  I'll work on this today(issue https://github.com/envoyproxy/envoy-perf/issues/90) and update in a subsequent PR.

File: salvo/src/lib/source_tree.py [hunk 9]
```diff
@@ -108,15 +123,33 @@ def get_origin(self) -> str:
...
+    log.debug(f"Valid: {valid} for object: {self}")
+
+    origin_url = self._source_url
+
+    cmd = "git remote -v"
+    if not origin_url:
+      cmd_params = cmd_exec.CommandParameters(cwd=self._source_path)
+      output = cmd_exec.run_command(cmd, cmd_params)
+      for line in output.split('\n'):
+        match = re.match(r'^origin\s*([\w:@\.\/-]+)\s\(fetch\)$', line)
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/source_tree.py:136]: Updated

File: salvo/src/lib/common/fileops.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/common/fileops.py:68]: I spent the day working on this and had to punt.  It's not going to land in this PR.

File: salvo/src/lib/source_tree.py [hunk 10]
```diff
@@ -40,24 +40,32 @@ def __init__(self, source_repo: proto_source.SourceRepository) -> None:
...
+      home_dir = constants.SALVO_TMP
+
+    self._tempdir = fileops.get_random_dir(home_dir)
 
     self._source_repo = source_repo
-    self._source_url = source_repo.source_url
-    self._branch = source_repo.branch
-    self._commit_hash = source_repo.commit_hash
-    self._source_path = source_repo.source_path
+    self._source_url = self._source_repo.source_url
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/source_tree.py:54]: Let me work on this.  I had this in a previous implementation but ended up changing it back.  I'll update.

File: salvo/src/lib/test_source_tree.py (hunk 1 above)
Comment (abaptiste) [reply salvo/src/lib/test_source_tree.py:17]: Fixed

File: salvo/src/lib/test_source_tree.py (hunk 3 above)
Comment (abaptiste) [reply salvo/src/lib/test_source_tree.py:243]: Guess I got a little happy witht he mocks.  I've removed `_check_output_side_effect_fail` and used `return_value` in this case.  I also consoliated the side effect functions as much as I could.

File: salvo/src/lib/test_source_tree.py (hunk 4 above)
Comment (abaptiste) [reply salvo/src/lib/test_source_tree.py:60]: Updated

File: salvo/src/lib/test_source_tree.py (hunk 5 above)
Comment (abaptiste) [reply salvo/src/lib/test_source_tree.py:363]: Updated

File: salvo/src/lib/source_tree.py [hunk 11]
```diff
@@ -158,6 +251,33 @@ def get_previous_commit_hash(self, current_commit: str,
     Returns:
       a string with the discovered commit hash
     """
+
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/source_tree.py:254]: I believe validate gets called at some point in the call tree. I just made a pass to verify.  If there's something I am still missing, please highlight.

File: salvo/src/lib/source_tree.py [hunk 12]
```diff
@@ -170,23 +290,57 @@ def get_revs_behind_parent_branch(self) -> int:
...
+    if not self._source_path:
+      raise SourceTreeError("Source is not checked out on disk")
+
+    cmd = "git status"
+    cmd_params = cmd_exec.CommandParameters(cwd=self._source_path)
+    status_output = cmd_exec.run_command(cmd, cmd_params)
+
+    commit_count = 0
+    ahead = re.compile(r'.*ahead of \'(.*)\' by (\d+) commit[s]')
+    uptodate = re.compile(r'Your branch is up to date with \'(.*)\'')
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/source_tree.py:304]: Fixed

File: salvo/src/lib/source_tree.py [hunk 13]
```diff
@@ -170,23 +290,57 @@ def get_revs_behind_parent_branch(self) -> int:
...
+
+    if not self._source_path:
+      raise SourceTreeError("Source is not checked out on disk")
+
+    cmd = "git status"
+    cmd_params = cmd_exec.CommandParameters(cwd=self._source_path)
+    status_output = cmd_exec.run_command(cmd, cmd_params)
+
+    commit_count = 0
+    ahead = re.compile(r'.*ahead of \'(.*)\' by (\d+) commit[s]')
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/source_tree.py:303]: Updated

File: salvo/src/lib/source_tree.py [hunk 14]
```diff
@@ -108,15 +123,33 @@ def get_origin(self) -> str:
...
+    valid = self._validate()
+    log.debug(f"Valid: {valid} for object: {self}")
+
+    origin_url = self._source_url
+
+    cmd = "git remote -v"
+    if not origin_url:
+      cmd_params = cmd_exec.CommandParameters(cwd=self._source_path)
+      output = cmd_exec.run_command(cmd, cmd_params)
+      for line in output.split('\n'):
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/source_tree.py:135]: I will add a block to raise an exception if the url is still empty.

File: salvo/src/lib/source_tree.py [hunk 15]
```diff
@@ -127,15 +160,69 @@ def pull(self) -> bool:
     Returns:
       a boolean indicating whether the operation was successful
     """
-    return False
+
+    source_name = proto_source.SourceRepository.SourceIdentity.Name(
+        self._source_repo.identity)
+    log.debug(f"Pulling [{source_name}] from origin: [{self._source_url}]")
+    self._validate()
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply salvo/src/lib/source_tree.py:167]: I can move the validate call, though I don't think that getting the source identity name from the value would affect things.

**Comment by abaptiste** [comment]
@mum4k This is ready for another pass from you, at your leisure.  Thanks!

File: salvo/src/lib/common/file_ops.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
+"""
```
Comment (mum4k) [inline salvo/src/lib/common/file_ops.py:1]: Sorry I must have missed this - can we also add tests for the `file_ops` package?

File: salvo/src/lib/common/fileops.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mum4k) [reply salvo/src/lib/common/fileops.py]: Good choice punting this and solving it outside of this PR. Thank you for filing the issue.

File: salvo/src/lib/source_tree.py [hunk 17]
```diff
@@ -108,15 +128,40 @@ def get_origin(self) -> str:
...
+      output = cmd_exec.run_command(cmd, cmd_params)
+      for line in output.split('\n'):
+        match = re.match(_GIT_ORIGIN_REGEX, line)
+        if match:
+          origin_url = match.group(1)
+          self._source_repo.source_url = origin_url
+          break
+
+    if not origin_url:
+      raise SourceTreeError(
```
Comment (mum4k) [inline salvo/src/lib/source_tree.py:149]: Can we add the `Raises:` section into the docstring now that the method can raise an exception?

File: salvo/src/lib/source_tree.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mum4k) [reply salvo/src/lib/source_tree.py]: Understood, this was mostly a question because the unusual placement (compared to the other methods) implied that this has some semantic meaning. Thanks for unifying it.

File: salvo/src/lib/source_tree.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mum4k) [reply salvo/src/lib/source_tree.py:265]: Thanks for clarifying.

File: salvo/src/lib/test_source_tree.py [hunk 18]
```diff
@@ -9,135 +9,428 @@
 from src.lib import (cmd_exec, source_tree)
 import api.source_pb2 as proto_source
 
+logging.basicConfig(level=logging.DEBUG)
 log = logging.getLogger(__name__)
 
+_GIT_TAG_LIST = """
```
Comment (mum4k) [inline salvo/src/lib/test_source_tree.py:15]: Moving the tag list into a global constant may be a regression. It introduces two problems:

1. The test cases are no longer "complete" or "all in one place" for the reader. The reader has to scroll up and down to understand them.
2. Multiple test cases use the same constant which to the reader implies that they all have to use the same constant / tag list. This likely isn't true and may confuse future maintainers.

How would you feel about moving this back into individual lists in the test cases that need this?

File: salvo/src/lib/test_source_tree.py [hunk 19]
```diff
@@ -9,135 +9,428 @@
...
+v1.14.5
```
Comment (mum4k) [inline salvo/src/lib/test_source_tree.py:16]: Thank you for making the list shorter. This is still bit longer than I expected, indicating that I might be missing something about the test cases. Following the [Zero one infinity rule](https://en.wikipedia.org/wiki/Zero_one_infinity_rule), do we need more than one or two tags to test the behavior we are after? Any more than the bare minimum makes it unclear as to what the tests depend on / really need.

**Review by mum4k — CHANGES_REQUESTED** [review]
Thank you for addressing the comments @abaptiste, sending the last batch.

File: salvo/src/lib/test_source_tree.py (hunk 19 above)
Comment (abaptiste) [reply salvo/src/lib/test_source_tree.py:16]: I wanted more than 2 tags in the list to verify we can discover an arbitrary number of commits backwards from a given hash.  Originally the intent was that we would test the hash N and N-1, but it maybe useful to test N-3 etc.  Having more than 2 tags in the list gives us the ability to verify this.

That said, I shortened the lists where applicable.

File: salvo/src/lib/test_source_tree.py (hunk 18 above)
Comment (abaptiste) [reply salvo/src/lib/test_source_tree.py:15]: Moved them back :)

File: salvo/src/lib/source_tree.py (hunk 17 above)
Comment (abaptiste) [reply salvo/src/lib/source_tree.py:149]: Certainly.  Updated.

File: salvo/src/lib/common/file_ops.py (hunk 16 above)
Comment (abaptiste) [reply salvo/src/lib/common/file_ops.py:1]: Added a test for file_ops

**Comment by abaptiste** [comment]
@mum4k Sorry for the delay.  I need a clone.  This is ready for another pass from you.

File: salvo/src/lib/common/test_file_ops.py [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+import os
+import json
+import tempfile
+
+from src.lib.common import file_ops
+
+from unittest import mock
+
+
+_test_json = """
```
Comment (mum4k) [inline salvo/src/lib/common/test_file_ops.py:16]: (nit) Constants should use CAPS_WITH_UNDER in their names, so this and the following constant should be named:

_TEST_JSON
_TEST_YAML

https://github.com/google/styleguide/blob/gh-pages/pyguide.md#3164-guidelines-derived-from-guidos-recommendations

**Review by mum4k — APPROVED** [review]
Looks good, we have one naming nit, please feel free to indicate if you would like to address it in a follow-up PR.

**Comment by mum4k** [comment]
No worries @abaptiste, there is no time pressure from my end. However when you do figure out how to get a clone, do let me know.

File: salvo/src/lib/common/test_file_ops.py (hunk 20 above)
Comment (abaptiste) [reply salvo/src/lib/common/test_file_ops.py]: I'll take care of it here.

**Comment by abaptiste** [comment]
@mum4k I addressed the last naming change.  Please let me know if you see anything else.  Thanks!

**Review by mum4k — APPROVED** [review]
Looks good, thank you @abaptiste and have a nice weekend!
